File: python/z3t_ai_agent/agent.py
from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import Any, Awaitable, Callable

import httpx

from .connection import Connection
from .context import CallContext, Send, create_call_context
from .llm import create_llm_clients
from .schema import VersionSchema
from .types import DEFAULTS, Logger, ResolvedConfig

Handler = Callable[[Any, CallContext], Awaitable[Any]]


@dataclass
class _QueuedCall:
    call_id: str
    schema_version: int
    input: Any
    send: Send


class Agent:
    def __init__(
        self,
        api_key: str,
        *,
        base_url: str | None = None,
        relay_urls: list[str] | None = None,
        timeout: float | None = None,
        max_concurrent_calls: int | None = None,
        reconnect_delay: float | None = None,
        max_reconnect_delay: float | None = None,
        logger: Logger | None = None,
    ) -> None:
        from .types import ConsoleLogger

        self._config = ResolvedConfig(
            api_key=api_key,
            base_url=base_url or DEFAULTS.base_url,
            # Developer-provided relay URLs take precedence — useful for local dev and
            # tests. Left empty, they're populated on start() via bootstrap.
            relay_urls=list(relay_urls) if relay_urls else [],
            timeout=timeout if timeout is not None else DEFAULTS.timeout,
            max_concurrent_calls=max_concurrent_calls if max_concurrent_calls is not None else DEFAULTS.max_concurrent_calls,
            reconnect_delay=reconnect_delay if reconnect_delay is not None else DEFAULTS.reconnect_delay,
            max_reconnect_delay=max_reconnect_delay if max_reconnect_delay is not None else DEFAULTS.max_reconnect_delay,
            logger=logger or ConsoleLogger(),
        )
        self._handlers: dict[int | str, Handler] = {}
        self._version_schemas: dict[int, VersionSchema] = {}
        self._connections: list[Connection] = []
        self._active_count = 0
        self._queue: list[_QueuedCall] = []
        self._http: httpx.AsyncClient | None = None

    def handle(
        self, version: int | None = None, schema: VersionSchema | None = None
    ) -> Callable[[Handler], Handler]:
        """Register a handler. Use as a decorator:

            @agent.handle()                          # default — all schema versions
            @agent.handle(version=1)                  # version-specific, no schema
            @agent.handle(version=1, schema=...)       # version-specific, typed schema

        The schema is synced with the platform on agent.start() and drives frontend
        form rendering and output display. Schemas sync as status='draft' by default —
        mutable, invisible to consumers, safe to keep editing across restarts. Set
        status='active' on the VersionSchema once ready to publish; from then on the
        schema is immutable and changing it will fail schema-sync.
        """
        if schema is not None and version is None:
            raise ValueError("a schema requires an explicit version")

        def decorator(fn: Handler) -> Handler:
            if version is None:
                self._handlers["default"] = fn
            else:
                self._handlers[version] = fn
                if schema is not None:
                    self._version_schemas[version] = schema
            return fn

        return decorator

    async def start(self) -> None:
        """Connect to the platform relay and begin handling calls. Runs until `stop()`
        is called (or an unhandled startup error occurs) — typically the last call in
        your program, e.g. `asyncio.run(agent.start())`.

        On startup, this:
        1. Fetches relay WebSocket URLs from the platform (unless overridden in config)
        2. Syncs any declared schemas (creates new versions as draft by default, deprecates removed ones)
        3. Opens a persistent WebSocket connection to each relay URL, and blocks until stopped

        Errors during bootstrap or schema sync are logged and abort startup.
        """
        self._http = httpx.AsyncClient()
        try:
            relay_urls = await self._bootstrap()
            if self._version_schemas:
                await self._sync_schemas()
        except Exception as exc:
            self._config.logger.error("[z3t SDK] Startup failed:", str(exc))
            await self._http.aclose()
            self._http = None
            return

        supported_versions = [v for v in self._handlers if isinstance(v, int)]
        self._connections = [
            Connection(url, self._config, self._dispatch, supported_versions) for url in relay_urls
        ]
        try:
            await asyncio.gather(*(conn.run() for conn in self._connections))
        finally:
            if self._http is not None:
                await self._http.aclose()
                self._http = None

    async def stop(self) -> None:
        """Disconnect from all relays. Useful for testing or graceful shutdown."""
        await asyncio.gather(*(conn.stop() for conn in self._connections))
        self._connections.clear()

    # ─── Private ─────────────────────────────────────────────────────────────

    async def _bootstrap(self) -> list[str]:
        if self._config.relay_urls:
            return self._config.relay_urls

        assert self._http is not None
        resp = await self._http.get(
            f"{self._config.base_url}/bootstrap",
            headers={"Authorization": f"Bearer {self._config.api_key}"},
        )
        if resp.is_error:
            raise RuntimeError(f"Bootstrap failed: HTTP {resp.status_code}")
        relay_urls = resp.json().get("relayUrls")
        if not relay_urls:
            raise RuntimeError("Bootstrap returned no relay URLs")
        return relay_urls

    async def _sync_schemas(self) -> None:
        versions = []
        for version, schema in self._version_schemas.items():
            entry: dict[str, Any] = {
                "version": version,
                "inputSchema": schema.input._def,
                "outputSchema": schema.output._def,
                "status": schema.status or "draft",
            }
            if schema.deprecates:
                entry["deprecates"] = schema.deprecates
            if schema.deprecation_notice:
                entry["deprecationNotice"] = schema.deprecation_notice
            versions.append(entry)

        assert self._http is not None
        resp = await self._http.post(
            f"{self._config.base_url}/schema-sync",
            json={"versions": versions},
            headers={"Authorization": f"Bearer {self._config.api_key}"},
        )
        if resp.is_error:
            raise RuntimeError(f"Schema sync failed: HTTP {resp.status_code}: {resp.text}")

        result = resp.json()
        deprecated = result.get("deprecatedVersions") or []
        if deprecated:
            self._config.logger.info(
                f"[z3t SDK] Schema versions deprecated: {', '.join(map(str, deprecated))}"
            )
        drafts = [v["version"] for v in (result.get("versions") or []) if v.get("status") == "draft"]
        if drafts:
            joined = ", v".join(map(str, drafts))
            self._config.logger.info(
                f"[z3t SDK] Synced as draft (not visible to consumers): v{joined} — "
                "set status='active' in .handle() to publish."
            )

    def _dispatch(self, call_id: str, schema_version: int, input: Any, send: Send) -> None:
        self._enqueue(_QueuedCall(call_id, schema_version, input, send))

    def _enqueue(self, call: _QueuedCall) -> None:
        if self._active_count < self._config.max_concurrent_calls:
            asyncio.create_task(self._process_call(call))
            return

        self._queue.append(call)

        max_queue = self._config.max_concurrent_calls * 2
        if len(self._queue) > max_queue:
            oldest = self._queue.pop(0)
            self._config.logger.warn(
                f"[z3t SDK] Queue depth exceeded (max {max_queue}) — rejecting call {oldest.call_id}"
            )
            asyncio.create_task(
                oldest.send({"type": "error", "callId": oldest.call_id, "message": "Queue depth exceeded"})
            )

    def _dequeue(self) -> None:
        if self._queue and self._active_count < self._config.max_concurrent_calls:
            asyncio.create_task(self._process_call(self._queue.pop(0)))

    async def _process_call(self, call: _QueuedCall) -> None:
        self._active_count += 1
        try:
            handler = self._handlers.get(call.schema_version) or self._handlers.get("default")
            if handler is None:
                await call.send(
                    {
                        "type": "error",
                        "callId": call.call_id,
                        "message": f"No handler for schema version {call.schema_version}",
                    }
                )
                return

            assert self._http is not None
            ctx = create_call_context(
                call.call_id,
                call.schema_version,
                call.send,
                self._config,
                create_llm_clients(self._config, call.call_id),
                self._http,
            )

            try:
                output = await asyncio.wait_for(handler(call.input, ctx), timeout=self._config.timeout)
                await call.send({"type": "result", "callId": call.call_id, "output": output})
            except asyncio.TimeoutError:
                await call.send({"type": "error", "callId": call.call_id, "message": "Handler timeout"})
            except Exception as exc:  # noqa: BLE001 — any handler exception becomes an error frame
                await call.send({"type": "error", "callId": call.call_id, "message": str(exc)})
        finally:
            self._active_count -= 1
            self._dequeue()

File: typescript/src/agent.ts
import { Connection, type CallDispatcher } from './connection'
import { createCallContext } from './context'
import { createLlmClients } from './llm'
import { DEFAULTS, type AgentConfig, type Handler, type ResolvedConfig, type WsSend } from './types'
import type { VersionSchema } from './schema'

interface QueuedCall {
  callId: string
  schemaVersion: number
  input: unknown
  send: WsSend
}

export class Agent {
  private readonly handlers = new Map<number | 'default', Handler>()
  private readonly versionSchemas = new Map<number, VersionSchema<unknown, unknown>>()
  private readonly config: ResolvedConfig
  private activeCount = 0
  private readonly queue: QueuedCall[] = []
  private readonly connections: Connection[] = []

  constructor(config: AgentConfig) {
    this.config = {
      baseUrl: DEFAULTS.baseUrl,
      timeout: DEFAULTS.timeout,
      maxConcurrentCalls: DEFAULTS.maxConcurrentCalls,
      reconnectDelay: DEFAULTS.reconnectDelay,
      maxReconnectDelay: DEFAULTS.maxReconnectDelay,
      logger: console,
      relayUrls: [], // populated on start() via bootstrap or config override
      ...config,
    }
  }

  /** Register a versioned handler with an input/output schema.
   *
   * The schema is synced with the platform on agent.start() and drives frontend
   * form rendering and output display. TypeScript infers the input/output types
   * from the schema so the handler is fully typed.
   *
   * Schemas sync as `status: 'draft'` by default — mutable, invisible to consumers,
   * safe to keep editing across restarts. Set `status: 'active'` once ready to publish;
   * from then on the schema is immutable (changing it will fail schema-sync).
   *
   * @example
   * agent.handle(1, {
   *   input: s.object({ doc: s.fileUri() }),
   *   output: s.object({ summary: s.markdown() }),
   *   status: 'active', // omit while iterating — defaults to 'draft'
   * }, async (input, ctx) => {
   *   // input.doc is typed as string
   * })
   */
  handle<I, O>(version: number, schema: VersionSchema<I, O>, handler: Handler<I, O>): this

  /** Register a versioned handler without an inline schema. Runs for that version but
   *  declares no schema itself — the version's schema must already exist (declared on a
   *  previous run, or by another handler). */
  handle(version: number, handler: Handler): this

  /** Register a default handler that runs for every schema version. Declares no schema
   *  itself — provide one via the versioned `handle(version, schema, handler)` overload. */
  handle(handler: Handler): this

  handle(
    versionOrHandler: number | Handler,
    schemaOrHandler?: VersionSchema<unknown, unknown> | Handler,
    handler?: Handler,
  ): this {
    if (typeof versionOrHandler === 'function') {
      this.handlers.set('default', versionOrHandler)
    } else if (typeof schemaOrHandler === 'function') {
      this.handlers.set(versionOrHandler, schemaOrHandler)
    } else if (schemaOrHandler && handler) {
      this.handlers.set(versionOrHandler, handler)
      this.versionSchemas.set(versionOrHandler, schemaOrHandler)
    }
    return this
  }

  /** Connect to the platform relay and begin handling calls.
   *
   * On first call, this:
   * 1. Fetches relay WebSocket URLs from the platform (unless overridden in config)
   * 2. Syncs any declared schemas (creates new versions as draft by default, deprecates removed ones)
   * 3. Opens a persistent WebSocket connection to each relay URL
   *
   * Errors during bootstrap or schema sync are logged and abort the startup.
   */
  start(): void {
    this.bootstrap()
      .then(({ relayUrls }) => {
        if (this.versionSchemas.size > 0) {
          return this.syncSchemas().then(() => relayUrls)
        }
        return relayUrls
      })
      .then((relayUrls) => this.connectAll(relayUrls))
      .catch((err: Error) => {
        this.config.logger.error('[z3t SDK] Startup failed:', err.message)
      })
  }

  /** Disconnect from all relays. Useful for testing or graceful shutdown. */
  stop(): void {
    for (const conn of this.connections) conn.stop()
    this.connections.length = 0
  }

  // ─── Private ─────────────────────────────────────────────────────────────

  private async bootstrap(): Promise<{ relayUrls: string[] }> {
    // Developer-provided relay URLs take precedence — useful for local dev and tests
    if (this.config.relayUrls.length > 0) {
      return { relayUrls: this.config.relayUrls }
    }

    const res = await fetch(`${this.config.baseUrl}/bootstrap`, {
      headers: { Authorization: `Bearer ${this.config.apiKey}` },
    })
    if (!res.ok) {
      throw new Error(`Bootstrap failed: HTTP ${res.status}`)
    }
    const { relayUrls } = (await res.json()) as { relayUrls: string[] }
    if (!relayUrls?.length) throw new Error('Bootstrap returned no relay URLs')
    return { relayUrls }
  }

  private async syncSchemas(): Promise<void> {
    const versions = [...this.versionSchemas.entries()].map(([version, schema]) => ({
      version,
      inputSchema: schema.input._def,
      outputSchema: schema.output._def,
      status: schema.status ?? 'draft',
      ...(schema.deprecates?.length ? { deprecates: schema.deprecates } : {}),
      ...(schema.deprecationNotice ? { deprecationNotice: schema.deprecationNotice } : {}),
    }))

    const res = await fetch(`${this.config.baseUrl}/schema-sync`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${this.config.apiKey}`,
      },
      body: JSON.stringify({ versions }),
    })

    if (!res.ok) {
      const body = await res.text().catch(() => '')
      throw new Error(`Schema sync failed: HTTP ${res.status}: ${body}`)
    }

    const result = (await res.json()) as {
      deprecatedVersions?: number[]
      versions?: Array<{ version: number; status: string }>
    }
    if (result.deprecatedVersions?.length) {
      this.config.logger.info(
        `[z3t SDK] Schema versions deprecated: ${result.deprecatedVersions.join(', ')}`,
      )
    }
    const drafts = result.versions?.filter((v) => v.status === 'draft').map((v) => v.version)
    if (drafts?.length) {
      this.config.logger.info(
        `[z3t SDK] Synced as draft (not visible to consumers): v${drafts.join(', v')} — ` +
          `set status: 'active' in .handle() to publish.`,
      )
    }
  }

  private connectAll(relayUrls: string[]): void {
    const supportedVersions = [...this.handlers.keys()].filter(
      (v): v is number => typeof v === 'number',
    )

    const dispatch: CallDispatcher = (callId, schemaVersion, input, send) => {
      this.enqueue({ callId, schemaVersion, input, send })
    }

    for (const url of relayUrls) {
      const conn = new Connection(url, this.config, dispatch, supportedVersions)
      this.connections.push(conn)
      conn.start()
    }
  }

  private enqueue(call: QueuedCall): void {
    if (this.activeCount < this.config.maxConcurrentCalls) {
      this.processCall(call)
      return
    }

    this.queue.push(call)

    const maxQueue = this.config.maxConcurrentCalls * 2
    if (this.queue.length > maxQueue) {
      const oldest = this.queue.shift()!
      this.config.logger.warn(
        `[z3t SDK] Queue depth exceeded (max ${maxQueue}) — rejecting call ${oldest.callId}`,
      )
      oldest.send({ type: 'error', callId: oldest.callId, message: 'Queue depth exceeded' })
    }
  }

  private dequeue(): void {
    if (this.queue.length > 0 && this.activeCount < this.config.maxConcurrentCalls) {
      this.processCall(this.queue.shift()!)
    }
  }

  private processCall(call: QueuedCall): void {
    this.activeCount++

    const handler = this.handlers.get(call.schemaVersion) ?? this.handlers.get('default')
    if (!handler) {
      this.activeCount--
      call.send({
        type: 'error',
        callId: call.callId,
        message: `No handler for schema version ${call.schemaVersion}`,
      })
      this.dequeue()
      return
    }

    const ctx = createCallContext(call.callId, call.schemaVersion, call.send, this.config, createLlmClients(this.config, call.callId))

    const handlerPromise = handler(call.input as Record<string, unknown>, ctx)
    const timeoutPromise = new Promise<never>((_, reject) => {
      setTimeout(() => reject(new Error('Handler timeout')), this.config.timeout)
    })

    Promise.race([handlerPromise, timeoutPromise])
      .then((output) => {
        call.send({ type: 'result', callId: call.callId, output })
      })
      .catch((err: Error) => {
        call.send({ type: 'error', callId: call.callId, message: err.message })
      })
      .finally(() => {
        this.activeCount--
        this.dequeue()
      })
  }
}
